diff --git a/src/Text/Pandoc/Writers/Haddock.hs b/src/Text/Pandoc/Writers/Haddock.hs
index 75b8c78dc..6cb720489 100644
--- a/src/Text/Pandoc/Writers/Haddock.hs
+++ b/src/Text/Pandoc/Writers/Haddock.hs
@@ -45,7 +45,6 @@ import Text.Pandoc.Options
 import Text.Pandoc.Pretty
 import Text.Pandoc.Shared
 import Text.Pandoc.Templates (renderTemplate')
-import Text.Pandoc.Writers.Math (texMathToInlines)
 import Text.Pandoc.Writers.Shared
 
 type Notes = [[Block]]
@@ -250,11 +249,10 @@ inlineToHaddock _ (Code _ str) =
   return $ "@" <> text (escapeString str) <> "@"
 inlineToHaddock _ (Str str) =
   return $ text $ escapeString str
-inlineToHaddock opts (Math mt str) = do
-  let adjust x = case mt of
-                      DisplayMath -> cr <> x <> cr
-                      InlineMath  -> x
-  adjust <$> (lift (texMathToInlines mt str) >>= inlineListToHaddock opts)
+inlineToHaddock _ (Math mt str) =
+  return $ case mt of
+    DisplayMath -> cr <> "\\[" <> text str <> "\\]" <> cr
+    InlineMath  -> "\\(" <> text str <> "\\)"
 inlineToHaddock _ il@(RawInline f str)
   | f == "haddock" = return $ text str
   | otherwise = do
diff --git a/test/writer.haddock b/test/writer.haddock
index 7f783abd1..5cd8fa9c2 100644
--- a/test/writer.haddock
+++ b/test/writer.haddock
@@ -454,15 +454,13 @@ ______________________________________________________________________________
 = LaTeX
 #latex#
 
--   
--   2 + 2 = 4
--   /x/ ∈ /y/
--   /α/ ∧ /ω/
--   223
--   /p/-Tree
--   Here’s some display math:
-    $$\\frac{d}{dx}f(x)=\\lim_{h\\to 0}\\frac{f(x+h)-f(x)}{h}$$
--   Here’s one that has a line break in it: /α/ + /ω/ × /x/2.
+-   \(2+2=4\)
+-   \(x \in y\)
+-   \(\alpha \wedge \omega\)
+-   \(223\)
+-   \(p\)-Tree
+    \[\frac{d}{dx}f(x)=\lim_{h\to 0}\frac{f(x+h)-f(x)}{h}\]
+-   Here’s one that has a line break in it: \(\alpha + \omega \times x^2\).
 
 These shouldn’t be math:
 
